### PR TITLE
Fix for issue 109

### PR DESCRIPTION
**Pull request by sband**

### CODE REVIEW

**Comment by sband** [comment]
This PR is not from my master, i will raise another one that is from my master, will avoid code not relevant to this fix

**Comment by sband** [comment]
I am not sure what is happening, why am i getting to see the code changes related to previous fixes i did in PR #112 . Anyways i am reopening this one again for review. I raised a PR from my master after merging code changes from sband:issue/109 but still it shows that my master is 8 commits behind the base master. 
All that aside, @Grundlefleck  Requesting your review comments on the fix i have made for this issue.
I haven't written any test cases yet, could not think of any which could fail this(or may be its just the distractions right now :-) ).

**Comment by Stephan202** [comment]
@sband try this:
```sh
git remote add upstream git@github.com:MutabilityDetector/MutabilityDetector.git
git fetch upstream
git checkout issue/109
git rebase upstream/master
git push --force-with-lease
```

I just checked out your fork, did the above (minus the push), and my `git log` now shows:
```
commit 35362500394073155f7364e996a5ba94f4c48f64 (HEAD -> issue/109)
Author: sband17 <band.sanket17@gmail.com>
Date:   Fri Mar 16 19:16:36 2018 +0530

    Fix for issue # 109

commit 1f277a2a1943bd95da572e9b07894a347ed4de1f (upstream/master)
Author: sband17 <band.sanket17@gmail.com>
Date:   Mon Mar 12 22:51:35 2018 +0530

    Improved fix for issue#111

commit b37f9d2d38ab06e3b95367f1f3fbc35f157a6c26
Author: sband17 <band.sanket17@gmail.com>
Date:   Sat Mar 10 14:56:23 2018 +0530

    Review comments given by @Grundlefleck implemented

commit de40a8d9ab5300299ac8b4ee4a57e5ad621203ba
Author: sband17 <band.sanket17@gmail.com>
Date:   Mon Mar 5 09:42:33 2018 +0530

    Fix for issue 111

commit 6ef5816ad6a04ff2d71b0ad6a101bfe02663005f
Author: Graham Allan <grundlefleck@gmail.com>
Date:   Wed Mar 7 23:38:26 2018 +0000

    Sign artifacts on deploy, instead of at verify.
    
    Is causing Travis build to fail, and is only needed when I do a release, not on every commit.

... etc ...
```

**Comment by sband** [comment]
@Stephan202  best !! that worked. Many thanks !

**Comment by sband** [comment]
@Grundlefleck  awaiting your comments on this fix

**Comment by Grundlefleck** [comment]
Sorry for the delay which I know must be frustrating. I will get to this, just not right now I'm afraid.

File: src/main/java/org/mutabilitydetector/checkers/AllChecksRunner.java [hunk 1]
```diff
@@ -62,6 +62,7 @@ public AnalysisResult runCheckers(ImmutableList<AnalysisResult> knownResultsSoFa
         Map<IsImmutable, Integer> results = newHashMap();
         Collection<MutableReasonDetail> reasons = newArrayList();
         Collection<AnalysisError> errors = newArrayList();
+        boolean isClassSelfReferenced=false;
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/AllChecksRunner.java]: Here and elsewhere: try to follow the existing code style with respect to whitespace: spaces around `=`, space after comma, space after `if`, etc.

File: src/main/java/org/mutabilitydetector/checkers/AsmMutabilityChecker.java [hunk 2]
```diff
@@ -42,11 +42,20 @@
...
     }
 
+    protected void setIsClassSelfReferenced(boolean isClassSeflReferenced) {
+        this.isClassSelfReferenced = isClassSeflReferenced;
+    }
+    
+    protected Boolean isClassSelfReferenced() {
+        return new Boolean(isClassSelfReferenced);
+    }
     protected void setResult(String message, CodeLocation<?> location, Reason reason) {
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/AsmMutabilityChecker.java]: There should be an empty line between these methods.

File: src/main/java/org/mutabilitydetector/checkers/CollectionWithMutableElementTypeToFieldChecker.java [hunk 3]
```diff
@@ -54,7 +54,7 @@
     
     private final Map<String, String> fieldSignatures = newHashMap();
     private final AnalysisInProgress analysisInProgress;
-
+
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/CollectionWithMutableElementTypeToFieldChecker.java]: Here and below: these unrelated whitespace changes can be reverted to minimize the diff.

File: src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java [hunk 4]
```diff
@@ -169,9 +174,8 @@ private void checkIfClassIsMutable(FieldInsnNode fieldInsnNode, Type typeAssigne
                     setAssigningToGenericFieldResult(fieldName, fieldLocation);
                     break;
                 }
-
-                MutabilityLookup mutabilityLookup = mutableTypeInfo.resultOf(dotted(ownerClass), assignedToField, analysisInProgress);
-
+                MutabilityLookup mutabilityLookup = mutableTypeInfo.resultOf(dotted(ownerClass), assignedToField,analysisInProgress);
+
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java]: This change can be reverted?

File: src/main/java/org/mutabilitydetector/checkers/AsmMutabilityChecker.java [hunk 5]
```diff
@@ -42,11 +42,20 @@
...
         return ownerClass;
     }
 
+    protected void setIsClassSelfReferenced(boolean isClassSeflReferenced) {
+        this.isClassSelfReferenced = isClassSeflReferenced;
+    }
+    
+    protected Boolean isClassSelfReferenced() {
+        return new Boolean(isClassSelfReferenced);
+    }
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/AsmMutabilityChecker.java]: The boxing doesn't seem necessary, thus:
```java
protected boolean isClassSelfReferenced() {
    return isClassSelfReferenced;
}
```
Idem further down.

File: src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java [hunk 6]
```diff
@@ -95,8 +96,12 @@ public void visitFormalTypeParameter(String name) {
     @Override
     public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
         if (signature == null) { return null ; }
+        if(signature.contains(this.ownerClass())) {
+            this.setIsClassSelfReferenced(true);
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java]: - No need for `this.` here.
- Without further knowing the code: are you sure the `.contains` check is robust for any kind of combination and class and method signature? (Taking into account pathological cases such as a two FQCNs where one is a suffix of the other.)

File: src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java (hunk 4 above)
Comment (sband) [reply src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java]: Nope this cant be reverted because of code changes happened above this in method visitField

File: src/main/java/org/mutabilitydetector/checkers/CollectionWithMutableElementTypeToFieldChecker.java (hunk 3 above)
Comment (sband) [reply src/main/java/org/mutabilitydetector/checkers/CollectionWithMutableElementTypeToFieldChecker.java]: There is constructor definition after line 56, that is why the white space added or i might have removed it and then added it again when i was doing a code. i will do something to avoid the whitespace changes.

File: src/main/java/org/mutabilitydetector/checkers/AllChecksRunner.java (hunk 1 above)
Comment (sband) [reply src/main/java/org/mutabilitydetector/checkers/AllChecksRunner.java]: Yep, done

File: src/main/java/org/mutabilitydetector/checkers/AsmMutabilityChecker.java (hunk 5 above)
Comment (sband) [reply src/main/java/org/mutabilitydetector/checkers/AsmMutabilityChecker.java]: done, yeah not sure what was i thinking when i was boxing it to Boolean object

File: src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java (hunk 6 above)
Comment (sband) [reply src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java]: Removed the "this" reference, but just for curiosity is there any impact of using it or just for the aesthetics ?
Nice 2nd point ! I tested the for inner class references for the field as well,  the contains check did not affect the behavior of the code. I mean the code gives correct results. so it is safe i believe.

File: src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java [hunk 7]
```diff
@@ -169,9 +174,9 @@ private void checkIfClassIsMutable(FieldInsnNode fieldInsnNode, Type typeAssigne
                     setAssigningToGenericFieldResult(fieldName, fieldLocation);
                     break;
                 }
-
-                MutabilityLookup mutabilityLookup = mutableTypeInfo.resultOf(dotted(ownerClass), assignedToField, analysisInProgress);
-
+                
+                MutabilityLookup mutabilityLookup = mutableTypeInfo.resultOf(dotted(ownerClass), assignedToField,analysisInProgress);
+
```
Comment (Stephan202) [inline src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java]: I'm really only seeing whitespace changes here (and in a bunch of other places still):smile:

File: src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java (hunk 6 above)
Comment (Stephan202) [reply src/main/java/org/mutabilitydetector/checkers/MutableTypeToFieldChecker.java]: Indeed, I suggested dropping `this.` only for consistency with the surrounding code. :)

**Comment by Grundlefleck** [comment]
Thanks @Stephan202 for helping with git issues and providing initial feedback. The suggestions were all good ones.

@sband I think you're going along the right lines, and I can see that it passes the test cases from the ClientOfMutabilityDetector project. I think there is potentially an issue in this code:
```
if(signature.contains(ownerClass())) {
            setIsClassSelfReferenced(true);
            return super.visitField(access, name, desc, signature, value);
        }
```

I think I'd like to see a comparison that is more precise. I don't have a test case to demonstrate a failing case but I have a suspicion that the logic as it is ("signature-contains-classname") is vulnerable to accidental matching if packages and classes clash in some way. This is a *blocker*, and even if you prove that issue doesn't exist, I'd like to see more test cases closer to the code in question.

In terms of style, I don't think it's necessary in this case to push fields and methods related to this change into the superclass `AsmMutabilityChecker`. While I get there may be some duplication across 2 of the subclasses, it seems too specific to only 2 subclasses to push up the inheritance hierarchy (where it doesn't apply to approx 10 other checkers). It's a judgment call, and there are no hard and fast rules, but I think here I'd rather see duplication than unnecessary inheritance. This is not a blocker, just an opportunity for improvement.

I also don't think I quite understand the code in `AllChecksRunner`, and the interaction between the two checkers here:

```
        for (AsmMutabilityChecker checker : checkers) {
            if(checker instanceof CollectionWithMutableElementTypeToFieldChecker) {
                checker.setIsClassSelfReferenced(isClassSelfReferenced);
            }
            CheckerResult checkerResult = checkerRunner.run(checker, toAnalyse, knownResultsSoFar);
            if(checker instanceof MutableTypeToFieldChecker) {
                isClassSelfReferenced = checker.isClassSelfReferenced();
            }
            results.put(checkerResult.isImmutable, getNewCount(results, checkerResult.isImmutable));
            reasons.addAll(checkerResult.reasons);
            errors.addAll(checkerResult.errors);
        }
```
At least in `CollectionWithMutableElementTypeToFieldChecker`, `if(isClassSelfReferenced())` I think could be replaced with something like `if (genericType.type.equals(ownerClass))`, without any need for the `isClassSelfReferenced` field in the superclass. Similarly in `MutableTypeToFieldChecker`, I'm surprised that once the check is made, it sets the field, rather than just moving on to the next field. If both of those were implemented as described, there would be no need for the field/method in `AsmMutabilityChecker`, and for `AllChecksRunner` to pick out the particular checker to set fields on them. For me, not understanding the logic is a *blocker* here, even though the code may be functionally correct.

However, you've done the work of investigating, so I don't want to just assume there is not a good reason for the way you've made the change. Perhaps you could explain what you were attempting with the `isClassSelfReferenced` field and method in the superclass, and what problem you were trying to solve?

Many thanks for your efforts thus far. Sorry about the delay, been out of town all week, finding time to sit with the code has been a struggle.

**Comment by sband** [comment]
Thanks for the comments @Grundlefleck. I will see how i can improve on signature-contains-classname logic and come up with some test cases around this. 
For the code in AllChecksRunner, the idea is to share the information about class being self referenced ( in generics angular brackets) between two checkers (in this case - CollectionWithMutableElementTypeToFieldChecker and  MutableTypeToFieldChecker) so that the wrongful conclusion of the class being non-final does not happen. I could not find a better way to share this information however i will try and eliminate the use of boolean here.

**Comment by Grundlefleck** [comment]
Hey @sband, I hope I haven't disheartened you with my feedback? Of course if you're busy with "real life" that's obviously more important, but I wanted to check you're not unhappy with the collaboration so far?

**Comment by sband** [comment]
Hi Allan
I am happy with the collaboration and the detailed comments I get. In fact
that is what I really long for. Just that I have been busy with some
personal stuff (family) ... Will be out till 8th may ... I will catch-up on
my pending work when I am back in my work mode.


Thanks
Sanket

**Comment by sband** [comment]
@Grundlefleck  i am back, will resume work on this one in coming days now

**Comment by sband** [comment]
@Grundlefleck  simplified the fix.. let me know your opinion on this one. i see that the case for wildcards gets handled as part of checking the generic parameters in CollectionWithMutableElementTypeToFieldChecker#anyGenericParameterTypesAreMutable

**Comment by Grundlefleck** [comment]
Looks a lot better @sband, thanks!

Could probably do with some more test cases in the main project, but I see that it passes the given test case in the Client project, so happy to merge. Good work!
